File: leetcode1/222.完全二叉树的节点个数.cpp
/*
 * @lc app=leetcode.cn id=222 lang=cpp
 *
 * [222] 完全二叉树的节点个数
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int getDepth(TreeNode* node, TreeNode* TreeNode::*child) {
        if (!node) {
            return 0;
        }
        int ans = 0;
        while (node) {
            ++ans;
            node = node->*child;
        }
        return ans;
    }

    int countNodes1(TreeNode* root) {
        if (!root) {
            return 0;
        }
        int ans = 1;
        int ll = 0, lr = 0, rl = 0, rr = 0;
        if (root->left) {
            ll = getDepth(root->left, &TreeNode::left);
            lr = getDepth(root->left, &TreeNode::right);
            if (ll == lr) {
                ans += (pow(2, ll) - 1);
                return ans + countNodes(root->right);
            }
        }
        if (root->right) {
            rl = getDepth(root->right, &TreeNode::left);
            rr = getDepth(root->right, &TreeNode::right);
            if (rl == rr) {
                ans += (pow(2, rl) - 1);
                return ans + countNodes(root->left);
            }
        }
        return ans;
    }

    int countNodes2(TreeNode* root) {
        if (!root) {
            return 0;
        }
        int l = getDepth(root, &TreeNode::left);
        int r = getDepth(root, &TreeNode::right);
        if (l == r) {
            return pow(2, l) - 1;
        }
        return 1 + countNodes2(root->left) + countNodes(root->right);
    }

    bool exist(TreeNode *node, int val, int width) {
        while (nullptr != node && width > 0) {
            int v = (val >> (width - 1)) & 0x1;
            if (v == 0) {
                node = node->left;
            } else {
                node = node->right;
            }
            --width;
        }
        return node != nullptr;
    }
    // 2分法；
    int countNodes(TreeNode *root) {
        if (nullptr == root) {
            return 0;
        }
        int depth = getDepth(root, &TreeNode::left);
        int low = 0;
        int upper = pow(2, depth - 1) - 1;
        int high = upper;
        while (low + 1 < high) {
            int mid = low + (high - low) / 2;
            if (exist(root, mid, depth - 1)) {
                low = mid;
            } else {
                high = mid;
            }
        }
        if (exist(root, high, depth - 1)) {
            return upper + 1 + high;
        }
        return upper + 1 + low;
    }
};
// @lc code=end

File: python/222.完全二叉树的节点个数.py
#
# @lc app=leetcode.cn id=222 lang=python3
#
# [222] 完全二叉树的节点个数
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
def treeHigh(root):
    h = 0
    while root:
        h += 1
        root = root.left
    return h


def treeRightHigh(root):
    h = 0
    while root:
        h += 1
        root = root.right
    return h



def exist(root, value, index):
    while root and index > 1:
        v = (1 << (index - 2))
        if value & v:
            value ^= v
            root = root.right
        else:
            root = root.left
        index -= 1
    return root is not None


class Solution:
    def countNodes1(self, root: Optional[TreeNode]) -> int:
        # 空
        if root is None:
            return 0
        high = treeHigh(root)
        left = 1 << (high - 1)
        right = (1 << high) - 1
        while left + 1 < right:
            mid = (left + right) // 2
            if exist(root, mid, high):
                left = mid
            else:
                right = mid
        if exist(root, right, high):
            return right
        return left

    def countNodes(self, root: Optional[TreeNode]) -> int:
        # 空
        if root is None:
            return 0
        left = treeHigh(root)
        right = treeRightHigh(root)
        if left == right:
            return (1 << left) - 1
        left_sum = self.countNodes(root.left)
        right_sum = self.countNodes(root.right)
        return left_sum + right_sum + 1
# @lc code=end

